Add main navigation block

AshGDS opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What / Why
- Add main navigation block
- https://trello.com/c/O79e9zUQ/47-build-main-navigation
- Wasn't sure if tests were needed as there's no model for this block.

matthillco commented on app/assets/stylesheets/views/_landing_page/main-navigation.scss:
```diff
@@ -0,0 +1,30 @@
+@import "govuk_publishing_components/individual_component_support";
+
+.app-b-main-navigation {
+  padding-top: govuk-spacing(3);
```
Small nitpick, you could use shorthand here for all the padding values, either old style with four values, or new style using `padding-block` and `padding-inline`. But happy to approve it as is if you're doing it this way for clearer readability. Same comment for margins at lines 18-19.

AshGDS replied:
Thanks Matt - yep I find being more verbose more readable, but can combine them into one line if preferred.

matthillco replied:
No worries, happy to leave these as is.

matthillco reviewed (commented):
Looks good overall, nice job. Just one comment about whether to use shorthand properties or not.

matthillco commented on app/assets/stylesheets/views/_landing_page/main-navigation.scss:
```diff
@@ -0,0 +1,30 @@
+
+  @include govuk-font(19, $weight: normal);
+
+  &--active {
```
I'm not a fan of this format, it makes it hard to search for full classnames. Might be better to expand this out to the full class name for readability and searchability.

AshGDS replied:
Thanks Matt, I have updated it to use the full classname.

andysellick commented on app/assets/stylesheets/views/_landing_page/main-navigation.scss:
```diff
@@ -0,0 +1,30 @@
+    display: block;
+  }
+
+  @include govuk-font(19, $weight: normal);
```
Nitpick: could this line move up to join the others above the media query?

AshGDS replied:
Thanks, done 👍

andysellick commented on app/assets/stylesheets/views/_landing_page/main-navigation.scss:
```diff
@@ -0,0 +1,30 @@
+  list-style: none;
+  display: inline-block;
+  margin-right: govuk-spacing(9);
+  margin-top: govuk-spacing(3);
```
Nitpick: the spacing around these elements is inconsistent - `.app-b-main-navigation` has different padding top and bottom, but then the spacing is evened out by having top margin on the list items. Is there a reason it's like this, or could we just do

```CSS
.app-b-main-navigation {
  padding: govuk-spacing(6) 0;
}

.app-b-main-navigation__listitem {
  // no margin top
}
```

AshGDS replied:
Good spot - this was actually intentional for future proofing. If the block ends up with enough links to cause overflow, the rows of links will need space between them, and we need to achieve that while still having the balance of overall top and bottom spacing, which is why I wrote it in this way.

<img width="766" alt="image" src="https://github.com/user-attachments/assets/2a20cf2e-672f-4998-af53-e36bdfae3b9b">

andysellick replied:
THAT - is nicely done sir. Excellent. I withdraw my comment 😁

AshGDS commented:
@andysellick @matthillco thanks both - should be ready for re-review 👍

andysellick commented on app/views/landing_page/blocks/_main_navigation.html.erb:
```diff
@@ -0,0 +1,22 @@
+        <% if link["active"] %>
+          <%= link["label"] %>
+        <% else %>
+          <a href="<%= link["link"]%>" class="govuk-link govuk-link--no-underline"><%= link["label"]%></a>
```
Nitpick: this might be cleaner using `link_to` (https://apidock.com/rails/ActionView/Helpers/UrlHelper/link_to) something like

```
<%= link_to link["label"], link["link"], class: "govuk-link govuk-link--no-underline" %>
```

andysellick commented on lib/data/landing_page_content_items/landing_page.yaml:
```diff
@@ -1,4 +1,15 @@
+- type: main_navigation
+  links:
+  - label: Lorem for Ipsum
+    link: /ipsum-1
```
Nitpick: obviously this is placeholder data, but I think `link` isn't needed for the active item?

andysellick commented on app/assets/stylesheets/views/_landing_page/main-navigation.scss:
```diff
@@ -0,0 +1,30 @@
+  @include govuk-font(19, $weight: normal);
+
+  @include govuk-media-query($until: "desktop") {
+    display: block;
```
Sorry, I've just noticed that the designs have the mobile version as what looks like a dropdown - might be worth checking with design on how this is supposed to work (in the design the active menu item is separate from the dropdown for some reason).

![Screenshot 2024-10-09 at 12 52 12](https://github.com/user-attachments/assets/f0f76868-75c8-4d84-be3b-e2b03eb0f4ef)

andysellick replied:
Obviously this makes this block much more complicated, apologies.

AshGDS replied:
@andysellick No worries - thanks for spotting this. I think it's probably easier to have two blocks here, one being the existing desktop/tablet simple menu, and then separately I'll create the more complex mobile one, and show/hide each using our CSS breakpoints. Let me know if you disagree. Will investigate and update this PR once done.

andysellick replied:
I get that duplicating it might be easier but duplicating stuff makes for more stuff to maintain. If you can keep it as one thing I think that'd be better, but obviously if it's too hard for now we could duplicate and fix later.

Starting to think we should have a list of tech debt...

AshGDS commented:
Thanks @andysellick - I've added a mobile menu into the existing block. I've also done the PR changes as requested. It's ready for a re review but I appreciate I've added quite a bit extra code, so no rush.

I'm also getting an error regarding the code coverage being too low now, though I'm not sure if I can write any more jasmine tests, so not sure what to do here. If you know how to resolve this, please let me know!

andysellick commented on app/views/landing_page/show.html.erb:
```diff
@@ -1,5 +1,5 @@
 
-<div class="landing-page">
+<div class="landing-page" id="content">
```
Thanks!

andysellick commented on app/assets/stylesheets/views/_landing_page/main-navigation.scss:
```diff
@@ -0,0 +1,89 @@
+  font-size: 15px;
+  top: 4px;
+  width: 20px;
+  content: "▼"
```
This element gets announced by Voiceover - ideally it wouldn't.

andysellick commented on app/assets/stylesheets/views/_landing_page/main-navigation.scss:
The mobile menu is touching the bottom of the container, should there be some padding here?

![Screenshot 2024-10-10 at 09 09 12](https://github.com/user-attachments/assets/9df70f37-2ef7-4724-8811-588a4f4169e8)

andysellick replied:
This now appears to be fixed 👍

andysellick commented on lib/data/landing_page_content_items/landing_page.yaml:
```diff
@@ -1,4 +1,21 @@
+  - label: Our Lorem
+    link: /our-lorem
+    active: false
+    children:
```
Having child menu items is really flexible, but is that a requirement? These child items aren't displayed on desktop either - how does that work? 

(sorry, feels like you've gone the extra mile here and I'm nitpicking)

AshGDS replied:
No worries @andysellick - The design seems to show child items on the mobile version, which don't exist on the desktop version. The implementation is quite simple, the child list elements just hide if the screen resolution is mobile. Do we need to ask someone about this functionality?

andysellick commented on app/assets/stylesheets/views/_landing_page/main-navigation.scss:
```diff
@@ -0,0 +1,89 @@
+@import "govuk_publishing_components/individual_component_support";
+
+
```
Nitpick: extra blank line.

andysellick commented on app/assets/stylesheets/views/_landing_page/main-navigation.scss:
```diff
@@ -0,0 +1,89 @@
+  @include govuk-media-query($until: "desktop") {
+    padding-top: 0;
+  }
+
```
Nitpick: empty line.

andysellick commented on app/assets/stylesheets/views/_landing_page/main-navigation.scss:
```diff
@@ -0,0 +1,89 @@
+  position: relative;
+  cursor: pointer;
+  color: govuk-colour("blue");
+
```
Nitpick: empty line.

andysellick commented on app/assets/stylesheets/views/_landing_page/main-navigation.scss:
```diff
@@ -0,0 +1,89 @@
+  @include govuk-media-query($until: "desktop") {
+    display: block;
+  }
+
```
Nitpick: empty line.

andysellick reviewed (commented):
Nice improvement, just a few comments.

AshGDS commented:
Thanks @andysellick - this should be ready for re-review. I'll try and get the coverage error resolved too.

AshGDS commented:
@andysellick The requirements for this have shifted a bit due to the design review, so probably worth holding off on the review for now 👍

AshGDS commented:
Thanks @andysellick - I've done the design refactor and pushed it as a separate commit, so it should be ready for another review.

<img width="1190" alt="image" src="https://github.com/user-attachments/assets/40c2d75c-a3e0-4bb0-ba26-2b0f0b5755ad">

<img width="473" alt="image" src="https://github.com/user-attachments/assets/74986559-85ea-4901-8caa-786ab73b0499">

andysellick commented on app/views/landing_page/blocks/_main_navigation.html.erb:
```diff
@@ -0,0 +1,46 @@
+          li_aria = { current: true } if link["active"]
+        %>
+        <%= tag.li(class: li_classes, aria: li_aria) do %>
+          <% if link["active"] %>
```
I think this bit down to around line 29 could be reduplicated a little, something like

```
          <%
            link_classes = "app-b-main-nav__link govuk-link govuk-link--no-underline"
            link_classes << " app-b-main-nav__link--active" if link["active"]
          %>
          <%= tag.a(href: link["link"], class: link_classes) do %>
            <span class="app-b-main-nav__mobile-border"></span>
            <%= link["label"] %>
          <% end %>
```

AshGDS replied:
Thanks, have cleaned up the code and fixed the bug, good spot 👍

andysellick reviewed (commented):
Looking good, just a few things that could be shorter.

andysellick commented on app/views/landing_page/blocks/_main_navigation.html.erb:
```diff
@@ -0,0 +1,46 @@
+%>
+
+<div class="app-b-main-nav" data-module="app-b-main-navigation">
+  <p class="app-b-main-nav__heading govuk-heading-s"><%= block.data["title"]%></p>
```
Quick one - was this supposed to be a link? I think that was raised last week.

AshGDS commented on app/assets/stylesheets/views/_landing_page/main-navigation.scss:
```diff
@@ -0,0 +1,153 @@
+  }
+}
+
+.js-app-b-main-nav__nav {
```
Thanks, renamed the class 👍

AshGDS commented:
@andysellick Thanks, it should be ready for review again (the last two commits.) I will ask the BE devs about the code coverage failures again in the tests.

andysellick reviewed (approved):
Code looks good to me, well done. I think there's still an outstanding question over that text being a link or not, but I think I'm happy to approve this now and we can raise another PR to make that change if we need to.
